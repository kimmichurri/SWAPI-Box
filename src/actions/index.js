export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})

export const updateData = (data) => ({
  type: 'UPDATE_DATA',
  data
})

export const currentCards = (data) => ({
  type: 'CURRENT_CARDS',
  data
})

export const isLoading = (boolean) => ({
  type: 'IS_LOADING',
  isLoading: boolean
})

