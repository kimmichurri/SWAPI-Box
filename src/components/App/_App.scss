.app {
  text-align: center;
}

.message-space {
  height: 50px;
}

.error-message {
  color: white;
}