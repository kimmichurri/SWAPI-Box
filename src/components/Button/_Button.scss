.card-category-buttons {
    background-color: $main-button-color-inactive;
    border-radius: 4px;
    color: #31394D;
    font-family: $main-font;
    margin: 20px;
    padding: 10px;
    width: 100px;
        &:hover {
            background-color: $main-button-color-active;
            box-shadow: $active-button-box-shadow;
            cursor: pointer;
            font-weight: 600;
        }
}