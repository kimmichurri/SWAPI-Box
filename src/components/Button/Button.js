import React, { Component } from 'react';
import { peopleCleaner, findHomeworld, findSpecies, findResidents, planetCleaner, vehicleCleaner } from '../Helper/cleaner';
import { connect } from 'react-redux';
import { hasError, updateData, currentCards, isLoading } from '../../actions/index';
import { fetchAnything } from '../Helper/fetchAnything';

class Button extends Component {
     constructor() {
         super();
     }


  findCategory = (retrievedData, categoryName) => {
    let existingCategory = retrievedData.find((card) => {
      return card.category === categoryName
    })
    return existingCategory
  }

  checkStateForExistingData = (retrievedData, categoryName) => {
    if(retrievedData === undefined) {
      return false
    } else if(this.findCategory(retrievedData, categoryName) === undefined) {
      return false
    } else {
      return true
    }
  }

  getPeople = async () => {
    let dataExists = this.checkStateForExistingData(this.props.retrievedData, 'people')
    if (dataExists) {
      let cachedPeople = this.props.retrievedData.filter((card) => {
        return card.category === 'people'
      })
      this.props.currentCards(cachedPeople)
    } else {
      try {
        this.props.isLoading(true)
        const url = 'https://swapi.co/api/people';
        const data = await fetchAnything(url)
        const allPeople = await findHomeworld(data.results)
        const withSpecies = await findSpecies(allPeople)
        const cleanPeopleData = peopleCleaner(withSpecies)
        this.props.updateData([...this.props.retrievedData, ...cleanPeopleData])
        this.props.currentCards(cleanPeopleData)
        this.props.isLoading(false)
      } catch(error) {
        this.props.hasError('There was a problem retrieving the data')
      }
    }
  }

  getPlanets = async () => {
    let dataExists = this.checkStateForExistingData(this.props.retrievedData, 'planets')
    if (dataExists) {
      let cachedPlanets = this.props.retrievedData.filter((card) => {
        return card.category === 'planets'
      })
      this.props.currentCards(cachedPlanets)
    } else {
      try {
        this.props.isLoading(true)
        const url = 'https://swapi.co/api/planets/'
        const data = await fetchAnything(url)
        const withResidents = await findResidents(data.results)
        const cleanPlanetData = planetCleaner(withResidents)
        this.props.updateData([...this.props.retrievedData, ...cleanPlanetData])
        this.props.currentCards(cleanPlanetData)
        this.props.isLoading(false)
      } catch(error) {
        this.props.hasError('There was a problem retrieving the data')
      }
    }
  }

  getVehicles = async () => {
    let dataExists = this.checkStateForExistingData(this.props.retrievedData, 'vehicles')
    if (dataExists) {
      let cachedVehicles = this.props.retrievedData.filter((card) => {
        return card.category === 'vehicles'
      })
      this.props.currentCards(cachedVehicles)
    } else {
      try {
        this.props.isLoading(true)
        const url = 'https://swapi.co/api/vehicles/'
        const data = await fetchAnything(url)
        const cleanVehicleData = await vehicleCleaner(data.results)
        this.props.updateData([...this.props.retrievedData, ...cleanVehicleData])
        this.props.currentCards(cleanVehicleData)
        this.props.isLoading(false)
      } catch(error) {
        this.props.hasError('There was a problem retrieving the data')
      }
    }
  }

     render() {
         return (
             <div>
                 <button onClick={this.getPeople} className="card-category-buttons people">people</button>
                 <button onClick={this.getPlanets} className="card-category-buttons planets">planets</button>
                 <button onClick={this.getVehicles} className="card-category-buttons vehicles">vehicles</button>
             </div>
         )

     }

}

export const mapStateToProps = (state) => ({
    errorStatus: state.errorStatus,
    retrievedData: state.retrievedData,
    selectedCards: state.currentCards,
    loading: state.loading
  })
  
  export const mapDispatchToProps = (dispatch) => ({
    hasError: (message) => dispatch(hasError(message)),
    updateData: (data) => dispatch(updateData(data)),
    currentCards: (data) => dispatch(currentCards(data)),
    isLoading: (boolean) => dispatch(isLoading(boolean))
  })
  
  export default connect(mapStateToProps, mapDispatchToProps)(Button);
