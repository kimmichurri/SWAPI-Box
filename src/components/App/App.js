import React, { Component } from 'react';
import './_App.scss';
import Header from '../Header/Header';
import Favorite from '../Favorite/Favorite';
import Button from '../Button/Button';
import Scroll from '../Scroll/Scroll';
import CardContainer from '../CardContainer/CardContainer';
import { peopleCleaner, findHomeworld, findSpecies, findResidents, planetCleaner, vehicleCleaner } from '../Helper/cleaner';
import LoadingRequest from '../LoadingRequest/LoadingRequest';

class App extends Component {
  constructor() {
    super();
    this.state = {
      openingFilm: {},
      selectedCards: [],
      favorites: 0,
      loading: false,
      errorStatus: ''
    }
  }

  fetchAnything = async (url) => { 
    try {
        const response = await fetch(url)
        return response.json()
    } catch (error) {
        // throw new Error('Response not okay')
        this.setState({
          errorStatus: 'Error fetching data'
        })
    }        
}

  randomFilmNumber = () => {
    return Math.floor((Math.random() * 7) + 1);
  }

  async componentDidMount() {
    const url = `https://swapi.co/api/films/${this.randomFilmNumber()}`;
    const selectedFilm = await this.fetchAnything(url)
    console.log(selectedFilm)
    this.firstFilm(selectedFilm)
  }

  firstFilm = (selectedFilm) => {
    this.setState({
      openingFilm: {
        title: selectedFilm.title, 
        date: selectedFilm.release_date,
        crawl: selectedFilm.opening_crawl
      }
    })
  }

  getPeople = () => { 
    this.setState({ loading: true })
    const url = 'https://swapi.co/api/people'
      return this.fetchAnything(url)
      .then(data => findHomeworld(data.results))
      .then(uniquePeople => findSpecies(uniquePeople))
      .then(getCleanedPeople => peopleCleaner(getCleanedPeople))
      .then(cleanPeople => this.setState({selectedCards: cleanPeople, loading: false}))
  }

  getPlanets = () => {
    this.setState({ loading: true })
    const url = 'https://swapi.co/api/planets/'
      return this.fetchAnything(url)
      .then(data => findResidents(data.results))
      .then(getCleanedPlanets => planetCleaner(getCleanedPlanets))
      .then(cleanPlanets => this.setState({selectedCards: cleanPlanets, loading: false}))
  }

  getVehicles = () => {
    this.setState({ loading: true })
    const url = 'https://swapi.co/api/vehicles/'
    return this.fetchAnything(url)
      .then(data => vehicleCleaner(data.results))
      .then(cleanVehicles => this.setState({selectedCards: cleanVehicles , loading: false}))
  }

  render() {
    const { openingFilm, selectedCards, favorites, loading } = this.state
    return (
      <div className="app">
          <Favorite favorites={favorites}/>
          <Header />
          <article className="message-space">
          {loading ? (
            <LoadingRequest />
          ) : (
            <Button getPeople={this.getPeople} getPlanets={this.getPlanets} getVehicles={this.getVehicles}/>
          )
          } 
          </article>
          {selectedCards.length ? (
            <CardContainer selectedCards={selectedCards}/>
          ) : (
            <Scroll openingFilm={openingFilm}/>
          )}
      </div>
    );
  }
}

export default App;
