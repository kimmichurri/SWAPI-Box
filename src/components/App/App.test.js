import React from 'react';
import App from './App';
import { mapDispatchToProps, mapStateToProps } from './App';
import Button from '../Button/Button';
import { loadOpeningFilm } from '../../actions'
import { shallow } from 'enzyme';

describe('App', () => {
  let wrapper;
  let mockErrorStatus;
  let mockRetrievedData;
  let mockCurrentCards;

    mockErrorStatus = 'error';
    mockRetrievedData = [{id: 1}]
    mockCurrentCards = [{card: 'name'}]
  
  beforeEach(() => {
    wrapper = shallow(<App errorStatus={mockErrorStatus} retrievedData={mockRetrievedData} currentCards={mockCurrentCards}/>)
  })

  it('should match the snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  describe('firstFilm', () => {
    it.skip('firstFilm should set openingFilm state with correct properties', () => {
      const mockSelectedFilm = {
        title: 'A New Hope',
        release_date: 1977,
        opening_crawl: 'some text'
      }
      wrapper.instance().firstFilm(mockSelectedFilm)
      expect(wrapper.instance().state).toEqual({
        title: 'A New Hope',
        date: 1977,
        crawl: 'some text'
      })
    })
  })

  it.skip('getPeople should be called on click', () => {
    const mockGetPeople = jest.fn();
    const wrapper = shallow(<Button getPeople={mockGetPeople}/>)
    wrapper.find('.people').simulate('click');
    expect(mockGetPeople).toHaveBeenCalled();
  })

  it.skip('getPlanets should be called on click', () => {
    const mockGetPlanets = jest.fn();
    const wrapper = shallow(<Button getPlanets={mockGetPlanets}/>)
    wrapper.find('.planets').simulate('click');
    expect(mockGetPlanets).toHaveBeenCalled;
  })

  it.skip('getVehicles should be called on click', () => {
    const mockGetVehicles = jest.fn();
    const wrapper = shallow(<Button getVehicles={mockGetVehicles}/>)
    wrapper.find('.vehicles').simulate('click');
    expect(mockGetVehicles).toHaveBeenCalled();
  })

  describe('mapDispatchToProps', () => {

    it('should provide a method to dispatch', () => {
      const mockFilm = { title: 'A New Hope', date: 1977, crawl: 'crawl' }
      const mockDispatch = jest.fn()
      const actionToDispatch = loadOpeningFilm(mockFilm)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.loadOpeningFilm(mockFilm)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

  describe('mapStateToProps', () => {
    it('should return an object with an errorStatus', () => {
      const message = 'error'
      const mockState = {
        errorStatus: message
      }
      const expected = {
        errorStatus: message
      }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })

})
