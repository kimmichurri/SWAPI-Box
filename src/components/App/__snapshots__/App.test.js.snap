// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Favorite
          favorites={0}
        />,
        <Header />,
        <article
          className="message-space"
        >
          <Button
            getPeople={[Function]}
            getPlanets={[Function]}
            getVehicles={[Function]}
          />
        </article>,
        <Scroll
          openingFilm={Object {}}
        />,
      ],
      "className": "app",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "favorites": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Button
            getPeople={[Function]}
            getPlanets={[Function]}
            getVehicles={[Function]}
          />,
          "className": "message-space",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "getPeople": [Function],
            "getPlanets": [Function],
            "getVehicles": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "article",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "openingFilm": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Favorite
            favorites={0}
          />,
          <Header />,
          <article
            className="message-space"
          >
            <Button
              getPeople={[Function]}
              getPlanets={[Function]}
              getVehicles={[Function]}
            />
          </article>,
          <Scroll
            openingFilm={Object {}}
          />,
        ],
        "className": "app",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "favorites": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Button
              getPeople={[Function]}
              getPlanets={[Function]}
              getVehicles={[Function]}
            />,
            "className": "message-space",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "getPeople": [Function],
              "getPlanets": [Function],
              "getVehicles": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "article",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "openingFilm": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
